### ISSUE TITLE
iocpreactor socket.error: [Errno 10014] The system detected an invalid pointer address in attempting to use a pointer argument in a call

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4669]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#4669|
|Type|defect|
|Created|2010-10-05 08:54:17Z|
|Branch|https://github.com/twisted/twisted/tree/iocp-win64-4669|

This reminds me of [#10921](https://github.com/twisted/twisted/issues/10921), but I'm not sure it's related.

```
Traceback (most recent call last):
  File "C:\Python27\lib\site-packages\twisted\python\log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "C:\Python27\lib\site-packages\twisted\python\log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "C:\Python27\lib\site-packages\twisted\python\context.py", line 59, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "C:\Python27\lib\site-packages\twisted\python\context.py", line 37, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "C:\Python27\lib\site-packages\twisted\internet\iocpreactor\reactor.py", line 114, in _callEventCallback
    evt.callback(rc, bytes, evt)
  File "C:\Python27\lib\site-packages\twisted\internet\iocpreactor\tcp.py", line 581, in cbAccept
    self.handleAccept(rc, evt)
  File "C:\Python27\lib\site-packages\twisted\internet\iocpreactor\tcp.py", line 598, in handleAccept
    struct.pack('I', self.socket.fileno()))
  File "C:\Python27\lib\socket.py", line 222, in meth
    return getattr(self._sock,name)(*args)
socket.error: [Errno 10014] The system detected an invalid pointer address in attempting to use a pointer argument in a call
```

Probably related, this an x86-64 build of Python:
```
Python 2.7 (r27:82525, Jul  4 2010, 07:43:08) [MSC v.1500 64 bit (AMD64)] on win32
```

<details><summary>Searchable metadata</summary>

```
trac-id__4669 4669
type__defect defect
reporter__ghazel ghazel
priority__normal normal
milestone__ 
branch__branches_iocp_win64_4669 branches/iocp-win64-4669
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__iocpreactor_win32 iocpreactor win32
time__1286268857000000 1286268857000000
changetime__1320195230000000 1320195230000000
version__None None
owner__exarkun exarkun
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @ghazel|
|-|-|

Any reproduction instructions?

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_2"></a>|@ghazel set owner to @glyph|
|-|-|

Just hit that code path. This does it:

```
from twisted.internet import iocpreactor
iocpreactor.install()
from twisted.internet import reactor, protocol

sf = protocol.ServerFactory()
sf.protocol = protocol.Protocol
reactor.listenTCP(10203, sf)

cf = protocol.ClientFactory()
cf.protocol = protocol.Protocol
reactor.connectTCP('localhost', 10203, cf)

reactor.run()
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to PenguinOfDoom|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="makolb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|makolb commented|
|-|-|

I encountered this bug as well, and found a fix for it. In twisted/iocpreactor/tcp.py, there are two different lines that make the following call:
```
struct.pack('I', self.socket.fileno())
```
For 64-bit Windows, this call instead needs to be:
```
struct.pack('Q', self.socket.fileno())
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Cool!  It looks like there's enough information to write a unit test for this, and the patch should be simple.  One thing I'm not sure about yet is why none of the existing unit tests catch this.  It looks like the reproduction just sets up one TCP connection.  Lots of unit tests do that.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_8"></a>|@ghazel commented|
|-|-|

Is there a x86-64 test slave running an x86-64 build of Python? The unit tests do not pass for me - the first test_tcp test, twisted.test.test_tcp.test_buildProtocolClient, hangs forever.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Ah!  x86-64 Python, I see.  We do not have a slave running that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [33017]) Branching to 'iocp-win64-4669'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

We have a 64 bit Windows slave now.  [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/iocp-win64-4669)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @exarkun|
|-|-|

Looks good to me.  The builder is actually green, even.  And P is definitely the right value, not Q :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [33047]) Merge iocp-win64-4669

Author: exarkun
Reviewer: glyph
Fixes: [#4669](https://github.com/twisted/twisted/issues/4669)

Change IOCP reactor to pack socket descriptor numbers into a string of the
correct size for the platform - the size of a pointer, rather than the
size of an integer - so the setsockopt() call succeeds instead of failing
with an error about an invalid pointer.
